Adding dinamic array

BurundukA opened the pull request:
Adding dynamic.cpp, dynamic.h, test.cpp

BurundukA committed:
Adding dinamic array

Adding dynamic.cpp, dynamic.h, test.cpp

frogitcher commented on Bordukova_A_A/Dynamic_array/Dynamic.cpp:
```diff
@@ -0,0 +1,214 @@
+Dynamic_Array::Dynamic_Array()
+	: size(0)
+	, capacity(1)
+	, data(nullptr) {
```
Лучше выделять размера ноль, иначе можем удалять по нуллптр когда реаллокейтим

frogitcher replied:
И постоянно надо следить 
а не нуллптр ли?

frogitcher commented on Bordukova_A_A/Dynamic_array/Dynamic.cpp:
```diff
@@ -0,0 +1,214 @@
+	capacity(_size), 
+	data(nullptr){
+	if (size <= 0) {
+		throw std::invalid_argument("Size must be >0!");
```
Не будет, тк сайзт

frogitcher commented on Bordukova_A_A/Dynamic_array/Dynamic.cpp:
```diff
@@ -0,0 +1,214 @@
+	capacity(list.size()), 
+	data(nullptr){
+	data = new int[capacity];
+	if (size != 0) {
```
Не надо тк если ноль то ничего не ломается. Внутри сору цикл который просто ноль итераций сделает

frogitcher commented on Bordukova_A_A/Dynamic_array/Dynamic.cpp:
```diff
@@ -0,0 +1,214 @@
+	delete[] data;
+	size = 0;
+	capacity = 0;
+	data = nullptr;
```
Не надо. Достаточно размер обнулить. Пусть уж будет у нас выделенный кусок

frogitcher commented on Bordukova_A_A/Dynamic_array/Dynamic.cpp:
```diff
@@ -0,0 +1,214 @@
+	if (size == 0) {
+		throw std::underflow_error("Array is empty!");
+	}
+	if (index < 0 || index >= size) {
```
Не может меньше нуля тк сайзт

frogitcher commented on Bordukova_A_A/Dynamic_array/Dynamic.cpp:
```diff
@@ -0,0 +1,214 @@
+	int* new_data = new int[capacity];
+	std::copy(begin(), begin() + index, new_data);
+	std::copy(begin() + index + 1, end(), new_data + index);
+	delete[] data;
```
Не надо перевыдедять. Можно на имеющейся памяти сделать

frogitcher commented on Bordukova_A_A/Dynamic_array/Dynamic.cpp:
```diff
@@ -0,0 +1,214 @@
+void Dynamic_Array::insert(size_t index, int value)
+{
+	if (index >= size) {
+		throw std::out_of_range("Index is out of the range!");
```
Не. Только если больше сайз. Можешь в конец инсертить

ylivshits replied:
то есть если index==size, то делаем пушбэк

frogitcher commented on Bordukova_A_A/Dynamic_array/Dynamic.cpp:
```diff
@@ -0,0 +1,214 @@
+}
+void Dynamic_Array::assign(size_t new_size, int value)
+{
+	if (new_size < 0) {
```
Не может тк сайзт

frogitcher commented on Bordukova_A_A/Dynamic_array/Dynamic.cpp:
```diff
@@ -0,0 +1,214 @@
+}
+int* Dynamic_Array::end()
+{
+	return data + size  - 1;
```
Нет. Именно плюс сайз. Как в СТЛ

frogitcher replied:
Иначе например сортить не сможем

frogitcher replied:
И другие алгоритмы из СТл применять

frogitcher commented on Bordukova_A_A/Dynamic_array/Dynamic.cpp:
```diff
@@ -0,0 +1,214 @@
+				return false;
+			}
+		}
+		return true;
```
Можно свернуть через std:: equal

ylivshits replied:
можно все заменить на
return (size==other.size && std::equal(...))

frogitcher commented on Bordukova_A_A/Dynamic_array/Dynamic.cpp:
```diff
@@ -0,0 +1,214 @@
+	}
+	if (i <=0) {
+		throw std::length_error("Index must be > 0!");
+	}
```
Нет. Проверки хотим в эт. Тут только быстро элемент

frogitcher replied:
Поэтому например можно сделать этот метод inline

frogitcher commented on Bordukova_A_A/Dynamic_array/Dynamic.cpp:
```diff
@@ -0,0 +1,214 @@
+	int* new_data = new int[new_capacity];
+	std::copy(begin(), end(), new_data);
+	delete[] data;
+	data = new_data;
```
Хм. Есть же create_data, который везде юзается

frogitcher commented on Bordukova_A_A/Dynamic_array/Dynamic.h:
```diff
@@ -0,0 +1,52 @@
+	const size_t Size() const;
+	const size_t Capacity() const;
+	bool Empty() const;
+	int* Get_Data() const;
```
Можно эти четыре метода сделать inline

frogitcher commented on Bordukova_A_A/Dynamic_array/Dynamic.h:
```diff
@@ -0,0 +1,52 @@
+
+	//Pointers
+	int* begin();
+	int* end();
```
Можно их инлайн

frogitcher commented on Bordukova_A_A/Dynamic_array/test.cpp:
```diff
@@ -0,0 +1,97 @@
+    Dynamic_Array one(1, 12);
+    CHECK_THROWS_WITH(a.at(125), "Index is out of the range!");
+    CHECK(a.at(1) == 6);
+    CHECK(a[1] == 6);
```
Можно ещё проверить например, что работает стд сорт

BurundukA committed:
Correcting mistakes

Вроде всё исправила

ylivshits commented on Bordukova_A_A/Dynamic_array/Dynamic.h:
```diff
@@ -0,0 +1,73 @@
+	bool operator!=(const Dynamic_Array& other) const;
+
+	//Assignment
+	Dynamic_Array& operator=(const Dynamic_Array other);
```
...(const Dynamic_Array& other);

ylivshits commented on Bordukova_A_A/Dynamic_array/Dynamic.cpp:
```diff
@@ -0,0 +1,155 @@
+	}
+
+//Actions with arrays
+void Dynamic_Array::push_back(int value) {
```
можно реализовать через insert

ylivshits commented on Bordukova_A_A/Dynamic_array/Dynamic.cpp:
```diff
@@ -0,0 +1,155 @@
+void Dynamic_Array::clear()
+{
+	size = 0;
+	capacity = 0;
```
физическая память остается заполненной, так что обнулять capacity не нужно

ylivshits commented on Bordukova_A_A/Dynamic_array/Dynamic.cpp:
```diff
@@ -0,0 +1,155 @@
+	if (index < 0 || index >= size) {
+		throw std::out_of_range("Index is out of the range!");
+	}
+	for (size_t i = index; i < size - 1; ++i) {
+		data[i] = data[i + 1];
+	}
```
можно заменить на std::copy()

ylivshits commented on Bordukova_A_A/Dynamic_array/Dynamic.cpp:
```diff
@@ -0,0 +1,155 @@
+	}
+	else {
+		resize(size + 1);
+		for (int i = size - 1; i > index; --i) {
+			data[i] = data[i - 1];
+		}
```
можно заменить на std::copy_backward()

ylivshits commented on Bordukova_A_A/Dynamic_array/Dynamic.cpp:
```diff
@@ -0,0 +1,155 @@
+void Dynamic_Array::create_data(size_t new_size) {
+	int new_capacity = new_size * 2;
+	int* new_data = new int[new_capacity];
+	for (int i = 0; i < size; ++i) {
+		new_data[i] = data[i];
+	}
```
можно через std::copy()

ylivshits commented on Bordukova_A_A/Dynamic_array/Dynamic.h:
```diff
@@ -0,0 +1,73 @@
+	void swap(Dynamic_Array& other);
+
+	//Pointers
+	inline int* begin()
```
const

BurundukA committed:
Correcting mistakes

Correct mistakes

ylivshits commented on Bordukova_A_A/Dynamic_array/Dynamic.cpp:
```diff
@@ -50,7 +50,6 @@ void Dynamic_Array::pop_back()
-	capacity = 0;
 }
 void Dynamic_Array::erase(size_t index)
 {
```
index не может быть меньше нуля, т.к. size_t - беззнаковый тип

ylivshits commented on Bordukova_A_A/Dynamic_array/Dynamic.h:
```diff
@@ -40,11 +40,11 @@ class Dynamic_Array {
 
 	//Pointers
-	inline int* begin()
+	inline const int* begin()
```
inline const int* begin() const{}
или
inline int* begin() const{}
аналогично для end()

BurundukA committed:
Merge pull request #1 from frogitcher/main

1

BurundukA committed:
Correcting 2 last mistakes

Я не могу сделать пр, пока мои изменения не одобрят все, кто до этого его запрещал

BurundukA committed:
Merge branch 'Bordukova_An_a/Dynamic_array' of https://github.com/BurundukA/oop_misis_2022 into Bordukova_An_a/Dynamic_array
